sycl/test-e2e/Matrix/joint_matrix_down_convert_impl.hpp

```diff
@@ -0,0 +1,165 @@
+     auto accB = bufB.get_access<access::mode::read_write>(cgh);
+     auto accAC = bufAC.get_access<access::mode::write>(cgh);
+
+     cgh.parallel_for<class imatrix>(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [=](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]]
+
+         {
+           // The submatrix API has to be accessed by all the workitems in a
+           // subgroup these functions will be called once by the subgroup no
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<sub_group, bfloat16, use::a, TM, TK, layout::row_major>
+               sub_a;
+           joint_matrix<sub_group, bfloat16, use::a, TM, TK, layout::row_major>
+               sub_ac;
+           // For B, we assume B has been already VNNIed.
+           joint_matrix<sub_group, bfloat16, use::b, TK, TN,
+                        ext::intel::experimental::matrix::layout::packed>
+               sub_b;
+           joint_matrix<sub_group, float, use::accumulator, TM, TN> sub_c;
+
+           joint_matrix_load(
+               sg, sub_c,
+               accC.template get_multi_ptr<access::decorated::no>() +
+                   (sg_startx * TM) * N + sg_starty / SG_SZ * TN,
+               N, layout::row_major);
+           for (int k = 0; k < K / TK; k += 1) {
+             joint_matrix_load(
+                 sg, sub_a,
+                 accA.template get_multi_ptr<access::decorated::no>() +
+                     (sg_startx * TM) * K + k * TK,
+                 K);
+             joint_matrix_load(
+                 sg, sub_b,
+                 accB.template get_multi_ptr<access::decorated::no>() +
+                     (k * TK / 2) * (N * 2) + sg_starty / SG_SZ * TN * 2,
+                 N * 2);
+             sub_c = joint_matrix_mad(sg, sub_a, sub_b, sub_c);
+           }
+           // This will be replaced by joint_matrix_copy API
+           // joint_matrix_copy(sg, sub_c, sub_ac);
+           auto wi_slice_c =
+               sycl::ext::intel::experimental::matrix::get_wi_data(sg, sub_c);
+           auto wi_slice_ac =
+               sycl::ext::intel::experimental::matrix::get_wi_data(sg, sub_ac);
+           for (int i = 0; i < wi_slice_c.length(); i++) {
+             wi_slice_ac[i] = (bfloat16)wi_slice_c[i];
+           }
+           ext::intel::experimental::matrix::joint_matrix_store(
+               sg, sub_ac,
+               accAC.template get_multi_ptr<access::decorated::no>() +
+                   (sg_startx * TM) * N + sg_starty / SG_SZ * TN,
+               N);
+         }); // parallel for
+   }).wait();
+}
+
+static constexpr size_t MATRIX_M = TM * 2;
```

Comment:
Would be good to move this block of global data to functions level to make it local to avoid potential errors